### DISCUSSION TITLE
Improve Timer creation

### DISCUSSION THREAD

**Discussion by otiv-wannes-vanleemput** [slint-ui/slint#5030] [General]
Hi all!

I'm developing a real-time video stream application, which uses a combination of Rust, Slint and Gstreamer.
To ensure the video being real-time, I constructed a Timer which fires after x ms, with a callback that shuts down the application. This timer resets every time a GStreamer AppSink receives a new sample (frame). 
The main issue here is that the creation of the timers happens before the main application loop, but I want to define the callback on the timer at creation.

This is a sample of what the code could look like:
```
pub fn create_shutdown_timer(
    slint_application_weak_ref: &slint::Weak<ui::Application>,
    timout_secs: u64,
) -> Result<Arc<Mutex<Timer>>, Error> {
    let application_shutdown_timer = Arc::new(Mutex::new(Timer::default()));
    let app_weak_upgraded = ui::try_slint_app_weak_reference_upgrade(slint_application_weak_ref)?;
    {
        match application_shutdown_timer.lock() {
            Ok(application_shutdown_timer_lock) => {
                application_shutdown_timer_lock.start(
                    TimerMode::SingleShot,
                    std::time::Duration::from_secs(timout_secs),
                    {
                        move || {
                            app_weak_upgraded.invoke_stop_video_stream();
                        }
                    },
                );
                application_shutdown_timer_lock.stop();
            }
            Err(e) => eprintln!("Error starting application shutdown timer: {e:?}"),
        }
    }
    Ok(application_shutdown_timer)
}
```

I'm starting and then stopping the timer to be able to set the callback, is there a way to make this more efficient?
It would be nice if it were possible to add this callback on creation, and then being able to start the timer with the callback in place.

**Comment by ogoffart** [comment]
Looking at your code now, I realize that Timer is Send, but it shouldn't be.
Moving Timer across thread may result in panics or in the wrong timer being stopped.

> I'm starting and then stopping the timer to be able to set the callback, is there a way to make this more efficient?

You're right that this is not very pretty code and we should do better.
But I think from a performance point of view, this should have very little impact

**Reply by otiv-wannes-vanleemput** [reply]
Thanks for the reply!
So if I understand correctly, I shouldn't be using Timer like this?
Starting and stopping the timer immediately indeed works just fine, I was mostly wondering whether I was missing something.
